File: core_v3/tdx_zt_update.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
通达信涨停板块 + 龙虎榜前20 自动更新脚本（Python版）
数据源：同花顺 dataapi（涨停池）+ ClawHub LHB API（龙虎榜）
用法：python tdx_zt_update.py [YYYYMMDD]

板块定义：
 - 首板 (SB.blk) = 1连板
 - 2连板 (2LB.blk) = 2连板
 - 3连板 (3LB.blk) = 3连板
 - 4连板 (4LB.blk) = 4连板
 - 5板以上 (5BYS.blk) = 5+连板
 - 所有连板 (SYLB.blk) = 2+连板
 - 昨封板 (ZFB.blk) = 目标日期涨停股
 - 昨日 (ZS.blk) = 目标日期涨停股（备用）
 - 龙虎榜前20 (LHBQ20.blk) = 龙虎榜净买入额前20
"""

import sys
import os
import json
import http.client
import urllib.parse
from datetime import datetime, timedelta

# 设置 stdout 为 UTF-8，避免 Windows GBK 编码错误
try:
    sys.stdout.reconfigure(encoding='utf-8')
    sys.stderr.reconfigure(encoding='utf-8')
except AttributeError:
    pass  # Python < 3.7 不支持

# ── 读取配置 ──────────────────────────────────────────────────────────────────
def get_config():
    # 命令行参数：argv[1] = 日期，argv[2] = TDX blocknew 目录
    if len(sys.argv) > 2 and sys.argv[2]:
        return sys.argv[2]
    # 读取同目录下的 config.json
    config_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'config.json')
    if os.path.exists(config_file):
        try:
            with open(config_file, 'r', encoding='utf-8') as f:
                cfg = json.load(f)
                if cfg.get('blocknewDir'):
                    return cfg['blocknewDir']
        except Exception:
            pass
    # 默认值
    return 'C:/new_tdx/T0002/blocknew'

TDX_DIR = get_config()

# ── 工具函数 ──────────────────────────────────────────────────────────────────
def http_get(url, headers=None, timeout=15):
    """简单的 HTTP GET，返回 (status_code, body_str)"""
    try:
        parsed = urllib.parse.urlparse(url)
        host = parsed.hostname
        port = parsed.port or (443 if parsed.scheme == 'https' else 80)
        path = parsed.path + ('?' + parsed.query if parsed.query else '')
        
        if parsed.scheme == 'https':
            conn = http.client.HTTPSConnection(host, port, timeout=timeout)
        else:
            conn = http.client.HTTPConnection(host, port, timeout=timeout)
        
        req_headers = headers or {}
        if 'User-Agent' not in req_headers:
            req_headers['User-Agent'] = 'Mozilla/5.0'
        
        conn.request('GET', path, headers=req_headers)
        resp = conn.getresponse()
        body = resp.read().decode('utf-8', errors='replace')
        conn.close()
        return resp.status, body
    except Exception as e:
        print(f'  [ERROR] HTTP 请求失败: {e}')
        return 0, ''

def write_blk_file(filepath, stocks):
    """写入 .blk 文件，每行一个7位代码"""
    lines = []
    for s in stocks:
        code = str(s.get('code', '')).strip()
        # 补齐6位，加市场前缀
        code6 = code.zfill(6)
        # 判断市场：6开头=沪市(1)，0/3开头=深市(0)，8/4开头=北交所(3)
        if code6.startswith('6'):
            prefix = '1'
        elif code6.startswith(('0', '3')):
            prefix = '0'
        else:
            prefix = '3'
        lines.append(prefix + code6)
    
    content = '\n'.join(lines) + '\n'
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(content)
    print(f'  [OK] {os.path.basename(filepath)}: {len(lines)} 只')

def get_target_date():
    """获取目标日期（命令行参数或昨天）"""
    if len(sys.argv) > 1 and sys.argv[1]:
        return sys.argv[1]
    yesterday = datetime.now() - timedelta(days=1)
    return yesterday.strftime('%Y%m%d')

# ── 获取涨停数据（同花顺） ──────────────────────────────────────────────────
def fetch_limit_up_data(date_str):
    """
    获取指定日期的涨停数据
    返回：[{code, name, is_again_limit, change_rate}, ...]
    """
    url = f'https://data.10jqka.com.cn/dataapi/limit_up/limit_up_pool?date={date_str}&type=all&page=1&limit=200'
    headers = {
        'Referer': 'https://www.10jqka.com.cn/',
        'User-Agent': 'Mozilla/5.0',
    }
    status, body = http_get(url, headers)
    
    if status != 200 or not body:
        print(f'  [ERROR] 同花顺 API 请求失败 (status={status})')
        return []
    
    try:
        data = json.loads(body)
        if data.get('status_code') != 0:
            print(f'  [ERROR] 同花顺 API 返回错误: {data.get("status_msg")}')
            return []
        stocks = data.get('data', {}).get('info', [])
        print(f'  [OK] 获取到 {len(stocks)} 只涨停股 ({date_str})')
        return stocks
    except Exception as e:
        print(f'  [ERROR] 解析同花顺数据失败: {e}')
        return []

# ── 获取龙虎榜数据（ClawHub） ──────────────────────────────────────────────
def fetch_lhb_data(date_str):
    """
    获取龙虎榜数据
    返回：[{code, name, close, change, net_buy, reason}, ...] 按净买入额排序
    """
    url = f'http://fffy520.gicp.net:8003/api/lhb/daily?date={date_str}'
    status, body = http_get(url, timeout=15)
    
    if status != 200 or not body:
        print(f'  [ERROR] 龙虎榜 API 请求失败 (status={status})')
        return []
    
    try:
        data = json.loads(body)
        if data.get('code') == 200:
            lhb_list = data.get('data', [])
            # 按净买入额排序
            lhb_list.sort(key=lambda x: float(x.get('net_buy', 0)), reverse=True)
            print(f'  [OK] 获取到 {len(lhb_list)} 只龙虎榜股票')
            return lhb_list[:20]  # 只取前20
        else:
            print(f'  [ERROR] 龙虎榜 API 返回错误: {data.get("message")}')
            return []
    except Exception as e:
        print(f'  [ERROR] 解析龙虎榜数据失败: {e}')
        return []

# ── 计算连板天数 ──────────────────────────────────────────────────────────────
def calc_continuous_days(target_date, all_stocks_set):
    """
    计算每只股票的连板天数。
    每只股票独立向前递推，断了就停。
    只遍历已有数据的交易日，不跳过无数据的日期（防止周末真空导致连板虚增）。
    返回：{code: continuous_days}
    """
    target_stocks = all_stocks_set.get(target_date, [])
    if not target_stocks:
        return {}

    # 提取所有历史日期（不含目标日），按时间倒序排列
    hist_dates = sorted([d for d in all_stocks_set if d != target_date], reverse=True)

    # 预计算历史数据每天有哪些股票（加快查表）
    history = {}
    for d in hist_dates:
        history[d] = {str(s.get('code', '')).zfill(6) for s in all_stocks_set[d]}

    # 逐只股票独立计算连板天数
    continuous = {}
    for s in target_stocks:
        code = str(s.get('code', '')).zfill(6)
        days = 1  # 今天涨停至少算1天
        for d in hist_dates:
            prev_codes = history.get(d, set())
            if not prev_codes:
                continue  # 无数据的日期跳过（理论上不会触发）
            if code in prev_codes:
                days += 1
            else:
                break  # 连板断了，停止
        continuous[code] = days

    return continuous

# ── 主流程 ────────────────────────────────────────────────────────────────────
def main():
    target_date = get_target_date()
    print('=' * 60)
    print(f'通达信涨停板块更新 - {target_date}')
    print('=' * 60)
    
    # 1. 获取目标日期涨停数据
    print('\n【步骤1】获取涨停数据...')
    target_stocks = fetch_limit_up_data(target_date)
    
    if not target_stocks:
        print('[ERROR] 未获取到涨停数据，退出')
        return
    
    # 2. 获取历史数据（计算连板天数）
    print('\n【步骤2】获取历史数据（计算连板天数）...')
    all_stocks = {target_date: target_stocks}
    current = datetime.strptime(target_date, '%Y%m%d')
    for i in range(1, 6):  # 往前查5天
        check_date = (current - timedelta(days=i)).strftime('%Y%m%d')
        stocks = fetch_limit_up_data(check_date)
        if stocks:
            all_stocks[check_date] = stocks
    
    # 3. 计算连板天数
    print('\n【步骤3】计算连板天数...')
    continuous = calc_continuous_days(target_date, all_stocks)
    print(f'  连板股票数: {len(continuous)}')
    
    # 4. 分类
    print('\n【步骤4】分类写入板块...')
    sb = []    # 首板
    lb2 = []   # 2连板
    lb3 = []   # 3连板
    lb4 = []   # 4连板
    lb5 = []   # 5板以上
    sylb = []  # 所有连板
    
    for s in target_stocks:
        code = str(s.get('code', '')).zfill(6)
        days = continuous.get(code, 1)
        
        if days == 1:
            sb.append(s)
        elif days == 2:
            lb2.append(s)
            sylb.append(s)
        elif days == 3:
            lb3.append(s)
            sylb.append(s)
        elif days == 4:
            lb4.append(s)
            sylb.append(s)
        elif days >= 5:
            lb5.append(s)
            sylb.append(s)
    
    # 5. 写入 .blk 文件
    print('\n【步骤5】写入 .blk 文件...')
    write_blk_file(os.path.join(TDX_DIR, 'SB.blk'), sb)
    write_blk_file(os.path.join(TDX_DIR, '2LB.blk'), lb2)
    write_blk_file(os.path.join(TDX_DIR, '3LB.blk'), lb3)
    write_blk_file(os.path.join(TDX_DIR, '4LB.blk'), lb4)
    write_blk_file(os.path.join(TDX_DIR, '5BYS.blk'), lb5)
    write_blk_file(os.path.join(TDX_DIR, 'SYLB.blk'), sylb)
    write_blk_file(os.path.join(TDX_DIR, 'ZFB.blk'), target_stocks)
    write_blk_file(os.path.join(TDX_DIR, 'ZS.blk'), target_stocks)
    
    # 6. 获取龙虎榜数据
    print('\n【步骤6】获取龙虎榜前20...')
    lhb_stocks = fetch_lhb_data(target_date)
    if lhb_stocks:
        write_blk_file(os.path.join(TDX_DIR, 'LHBQ20.blk'), lhb_stocks)
    
    # 完成
    print('\n' + '=' * 60)
    print('✅ 更新完成！')
    print('=' * 60)
    print(f'\n板块文件已写入: {TDX_DIR}')
    print('请重启通达信查看效果。\n')

if __name__ == '__main__':
    main()

File: 连板龙AI自动复盘项目/core_v4/tdx_zt_update.js
#!/usr/bin/env node
/**
 * 通达信涨停板块 + 龙虎榜前20 自动更新脚本（完整版）
 * 数据源：同花顺 dataapi（涨停）+ ClawHub lhb-api（龙虎榜）
 * 用法：node tdx_zt_update.js [YYYYMMDD]
 *
 * 板块定义：
 *  - 首板 (SB.blk) = 1连板
 *  - 2连板 (2LB.blk) = 2连板
 *  - 3连板 (3LB.blk) = 3连板
 *  - 4连板 (4LB.blk) = 4连板
 *  - 5板以上 (5BYS.blk) = 5+连板
 *  - 所有连板 (SYLB.blk) = 2+连板
 *  - 昨封板 (ZFB.blk) = 目标日期涨停股
 *  - 龙虎榜前20 (LHBQ20.blk) = 龙虎榜净买入额前20
 */

const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');

// 读取配置：优先命令行参数，其次 config.json，最后默认值
function getConfig() {
    // 命令行参数：argv[3] = TDX目录
    if (process.argv[3]) {
        return process.argv[3];
    }
    // 读取同目录下的 config.json
    const configFile = path.join(path.dirname(process.argv[1]), 'config.json');
    if (fs.existsSync(configFile)) {
        try {
            const cfg = JSON.parse(fs.readFileSync(configFile, 'utf8'));
            if (cfg.blocknewDir) return cfg.blocknewDir;
        } catch (e) {}
    }
    // 默认值
    return 'C:/new_tdx/T0002/blocknew';
}

const TDX_DIR = getConfig();
const HISTORY_DAYS = 10; // 往前追溯10个交易日

// 获取上一个交易日
function getPrevTradingDay(dateStr) {
    const y = dateStr.slice(0, 4);
    const m = dateStr.slice(4, 6);
    const d = dateStr.slice(6, 8);
    const date = new Date(y + '-' + m + '-' + d);
    date.setDate(date.getDate() - 1);
    let day = date.getDay();
    if (day === 0) date.setDate(date.getDate() - 2);
    else if (day === 6) date.setDate(date.getDate() - 1);
    const mm = String(date.getMonth() + 1).padStart(2, '0');
    const dd = String(date.getDate()).padStart(2, '0');
    return date.getFullYear() + mm + dd;
}

// 获取目标日期（今天或指定日期）
function getTargetDate(dateArg) {
    if (dateArg) {
        return dateArg; // 指定日期模式：直接使用
    } else {
        let d = new Date();
        let day = d.getDay();
        if (day === 0) d.setDate(d.getDate() - 2);
        else if (day === 6) d.setDate(d.getDate() - 1);
        const hour = d.getHours();
        if (hour < 15) {
            d.setDate(d.getDate() - 1);
            day = d.getDay();
            if (day === 0) d.setDate(d.getDate() - 2);
            else if (day === 6) d.setDate(d.getDate() - 1);
        }
        const mm = String(d.getMonth() + 1).padStart(2, '0');
        const dd = String(d.getDate()).padStart(2, '0');
        return d.getFullYear() + mm + dd;
    }
}

// 获取交易日列表（从目标日期往前N天）
function getTradingDays(targetDate, days) {
    const dates = [targetDate];
    let prev = getPrevTradingDay(targetDate);
    for (let i = 0; i < days; i++) {
        dates.push(prev);
        prev = getPrevTradingDay(prev);
    }
    return dates;
}

// 从同花顺API获取涨停数据
function fetchLimitUpData(date) {
    return new Promise((resolve) => {
        const url = 'https://data.10jqka.com.cn/dataapi/limit_up/limit_up_pool?page=1&limit=200&field=199112,10,9001,330323,330324,330325,133971,9002,133970,1968584,3475914,9003&order_field=330324&order_type=0&date=' + date + '&_=' + Date.now();
        const req = https.get(url, {
            headers: {
                'User-Agent': 'Mozilla/5.0',
                'Referer': 'https://www.10jqka.com.cn/'
            }
        }, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    if (json.status_code === 0 && json.data && json.data.info) {
                        resolve(json.data.info);
                    } else {
                        resolve([]);
                    }
                } catch (e) {
                    resolve([]);
                }
            });
        });
        req.on('error', () => resolve([]));
        req.setTimeout(10000, () => {
            req.destroy();
            resolve([]);
        });
    });
}

// 从 ClawHub LHB API 获取龙虎榜数据
function fetchLHBData(date) {
    return new Promise((resolve) => {
        // date 格式：YYYYMMDD → YYYY-MM-DD
        const fmt = date.slice(0, 4) + '-' + date.slice(4, 6) + '-' + date.slice(6, 8);
        const url = 'http://fffy520.gicp.net:8003/api/lhb/daily?date=' + fmt;
        const req = http.get(url, (res) => {
            if (res.statusCode !== 200) {
                resolve([]);
                return;
            }
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    if (json.data && json.data.length > 0) {
                        resolve(json.data);
                    } else {
                        resolve([]);
                    }
                } catch (e) {
                    resolve([]);
                }
            });
        });
        req.on('error', () => resolve([]));
        req.setTimeout(10000, () => { req.destroy(); resolve([]); });
    });
}

// 计算连板天数
function calculateConsecutiveDays(targetDate, allData) {
    const targetStocks = allData[targetDate] || [];
    const consecutiveDays = {};

    for (const stock of targetStocks) {
        const code = stock.code;
        let days = 1;
        let checkDate = getPrevTradingDay(targetDate);

        while (allData[checkDate]) {
            const prevStocks = allData[checkDate];
            const found = prevStocks.find(s => s.code === code);
            if (found) {
                days++;
                checkDate = getPrevTradingDay(checkDate);
            } else {
                break;
            }
        }

        consecutiveDays[code] = days;
    }

    return consecutiveDays;
}

// 写入.blk文件
function writeBlkFile(filename, stocks) {
    const filepath = path.join(TDX_DIR, filename);
    const content = stocks.map(s => {
        const code = String(s.code).padStart(6, '0');
        let market = '0';
        if (code.startsWith('6')) market = '1';
        else if (code.startsWith('8') || code.startsWith('4')) market = '3';
        return market + code;
    }).join('\n') + '\n';
    fs.writeFileSync(filepath, content, 'utf8');
    console.log(`  ✓ ${filename}: ${stocks.length} 只`);
}

// 主函数
async function main() {
    const dateArg = process.argv[2];
    const targetDate = getTargetDate(dateArg);

    console.log('==================================================');
    console.log('通达信涨停板块 + 龙虎榜前20 自动更新脚本');
    console.log('==================================================');
    console.log(`目标日期: ${targetDate.slice(0,4)}-${targetDate.slice(4, 6)}-${targetDate.slice(6, 8)}`);
    console.log('');

    // 获取交易日列表
    console.log('【步骤1】获取交易日数据...');
    const tradingDays = getTradingDays(targetDate, HISTORY_DAYS);
    console.log(`  追溯 ${HISTORY_DAYS} 个交易日: ${tradingDays.slice(0, 6).join(', ')}...`);
    console.log('');

    // 获取所有交易日的数据
    console.log('【步骤2】获取涨停数据...');
    const allData = {};
    for (const date of tradingDays) {
        process.stdout.write(`  获取 ${date}... `);
        const stocks = await fetchLimitUpData(date);
        allData[date] = stocks;
        console.log(`${stocks.length} 只涨停`);
        await new Promise(resolve => setTimeout(resolve, 500));
    }
    console.log('');

    // 计算连板天数
    console.log('【步骤3】计算连板天数...');
    const consecutiveDays = calculateConsecutiveDays(targetDate, allData);
    const targetStocks = allData[targetDate] || [];
    console.log(`  ${targetDate} 涨停: ${targetStocks.length} 只`);
    console.log('');

    // 分类股票
    console.log('【步骤4】分类股票...');
    const categories = {
        shouban: [],    // 首板 (1连板)
        lianban2: [],    // 2连板
        lianban3: [],    // 3连板
        lianban4: [],    // 4连板
        lianban5: [],    // 5连板
        lianban5plus: [], // 6+连板
        allLianban: []   // 所有连板 (2+连板)
    };

    for (const stock of targetStocks) {
        const code = stock.code;
        const days = consecutiveDays[code] || 1;

        if (days === 1) {
            categories.shouban.push(stock);
        } else {
            categories.allLianban.push(stock);

            if (days === 2) categories.lianban2.push(stock);
            else if (days === 3) categories.lianban3.push(stock);
            else if (days === 4) categories.lianban4.push(stock);
            else if (days === 5) categories.lianban5.push(stock);
            else if (days >= 6) categories.lianban5plus.push(stock);
        }
    }

    console.log(`  首板: ${categories.shouban.length} 只`);
    console.log(`  2连板: ${categories.lianban2.length} 只`);
    console.log(`  3连板: ${categories.lianban3.length} 只`);
    console.log(`  4连板: ${categories.lianban4.length} 只`);
    console.log(`  5连板: ${categories.lianban5.length} 只`);
    console.log(`  6+连板: ${categories.lianban5plus.length} 只`);
    console.log(`  所有连板: ${categories.allLianban.length} 只`);
    console.log('');

    // 写入文件
    console.log('【步骤5】写入 .blk 文件...');
    writeBlkFile('SB.blk', categories.shouban);           // 首板
    writeBlkFile('SYLB.blk', categories.allLianban);     // 所有连板
    writeBlkFile('2LB.blk', categories.lianban2);       // 2连板
    writeBlkFile('3LB.blk', categories.lianban3);       // 3连板
    writeBlkFile('4LB.blk', categories.lianban4);       // 4连板

    // 5板以上 = 5连板 + 6+连板
    const lb5plus = [...categories.lianban5, ...categories.lianban5plus];
    writeBlkFile('5BYS.blk', lb5plus);                // 5板以上 (5+连板)
    writeBlkFile('ZFB.blk', targetStocks);                // 昨封板（目标日期涨停股）
    console.log('');

    // 龙虎榜前20
    console.log('【步骤6】龙虎榜前20...');
    const lhbData = await fetchLHBData(targetDate);
    if (lhbData && lhbData.length > 0) {
        // 按净买入额排序，取前20
        const top20 = [...lhbData].sort((a, b) => b.net_buy - a.net_buy).slice(0, 20);
        console.log(`  龙虎榜 ${lhbData.length} 只，取前20（按净买入额排序）:`);
        top20.forEach((s, i) => {
            console.log(`    ${i + 1}. ${s.code} ${s.name} 净买入:${s.net_buy}万 涨幅:${s.change}%`);
        });
        writeBlkFile('LHBQ20.blk', top20);
    } else {
        console.log('  龙虎榜数据获取失败');
    }
    console.log('');

    // 输出详细信息
    console.log('【复盘】各板块股票明细：');
    console.log('');

    if (categories.shouban.length > 0) {
        console.log(`首板(${categories.shouban.length}只):`);
        categories.shouban.forEach((s, i) => {
            console.log(`  ${i + 1}. ${s.code} ${s.name} 涨幅:${s.change_rate}%`);
        });
        console.log('');
    }

    if (categories.lianban2.length > 0) {
        console.log(`2连板(${categories.lianban2.length}只):`);
        categories.lianban2.forEach((s, i) => {
            console.log(`  ${i + 1}. ${s.code} ${s.name} 涨幅:${s.change_rate}%`);
        });
        console.log('');
    }

    if (categories.lianban3.length > 0) {
        console.log(`3连板(${categories.lianban3.length}只):`);
        categories.lianban3.forEach((s, i) => {
            console.log(`  ${i + 1}. ${s.code} ${s.name} 涨幅:${s.change_rate}%`);
        });
        console.log('');
    }

    if (categories.lianban4.length > 0) {
        console.log(`4连板(${categories.lianban4.length}只):`);
        categories.lianban4.forEach((s, i) => {
            console.log(`  ${i + 1}. ${s.code} ${s.name} 涨幅:${s.change_rate}%`);
        });
        console.log('');
    }

    if (categories.lianban5.length > 0) {
        console.log(`5连板(${categories.lianban5.length}只):`);
        categories.lianban5.forEach((s, i) => {
            console.log(`  ${i + 1}. ${s.code} ${s.name} 涨幅:${s.change_rate}%`);
        });
        console.log('');
    }

    if (categories.lianban5plus.length > 0) {
        console.log(`6+连板(${categories.lianban5plus.length}只):`);
        categories.lianban5plus.forEach((s, i) => {
            console.log(`  ${i + 1}. ${s.code} ${s.name} 涨幅:${s.change_rate}%`);
        });
        console.log('');
    }

    console.log('==================================================');
    console.log('✅ 完成！');
    console.log('==================================================');
}

main().catch(e => {
    console.error('错误:', e.message);
    process.exit(1);
});
